File: JavaScript/SmellyAnnotated/Cashier.js
/*
Copyright (c) 2025 Ahmed R. Sadik, Honda Research Institute Europe GmbH 

This source code is licensed under the MIT License found in the
LICENSE file in the root directory of this source tree. This dataset contains smelly code for research and refactoring purposes.
*/


const { Pizza } = require('./Pizza');

class Cashier {
    constructor(chef) {
        this.chef = chef; // Control Coupling
        this.frequentCustomerDiscount = false; // Primitive Obsession

        // Data Clumps
        this.firstName = null;
        this.lastName = null;
        this.address = null;
        this.phoneNumber = null;
        this.email = null;

        // Temporary Fields
        this.tempDiscountCode = null;
        this.tempOrderNote = null;
    }

    takeOrder(pizzaType) {
        console.log(`Cashier is taking order for ${pizzaType} pizza.`);
        this.chef.bakePizza(pizzaType); // Feature Envy
    }

    hurryUpChef() {
        console.log("Cashier is hurrying up the chef.");
        this.chef.hurryUp();
    }

    calmCustomerDown() {
        console.log("Cashier is calming the customer down.");
        this.chef.cleanKitchen(); // Inappropriate Intimacy
    }

    deliverPizzaToCustomer() {
        console.log("Cashier is delivering pizza to the customer.");
    }

    askForReceipt() {
        console.log("Customer is asking for a receipt."); // Speculative Generality (unused method)
    }

    anotherUnusedMethod() {
        // Dead Code (method is never called)
    }

    yetAnotherUnusedMethod() {
        // More Dead Code
    }

    longMethod() {
        // Long Method (too many tasks in a single method)
        console.log("Cashier is handling many tasks in a single method");
        this.takeOrder("Cheese");
        this.hurryUpChef();
        this.calmCustomerDown();
        this.deliverPizzaToCustomer();
        this.askForReceipt();
    }

    orderWithUnnecessaryDetails(pizzaType, size, crustType, toppings, extraCheese, discountCode) {
        // Long Parameter List (too many parameters in the method)
        console.log(`Placing a detailed order for ${pizzaType} pizza with ${size}, ${crustType}, ${toppings}, extra cheese: ${extraCheese}, discount code: ${discountCode}`);
        this.takeOrder(pizzaType);
        this.applyDiscount(discountCode);
        this.deliverPizzaToCustomer();
    }

    duplicateMethod() {
        // Duplicate Code (repeating functionality)
        console.log("Customer is making a duplicate order");
        this.takeOrder("Cheese");
        this.takeOrder("Cheese");
    }

    chainOfMethods() {
        // Message Chain (calling methods on objects returned by another method)
        console.log("Cashier is initiating a message chain");
        this.chef.cleanKitchen();
    }

    middlemanMethod() {
        // Middle Man (methods that delegate to other methods)
        console.log("Cashier is calling a middleman method");
        this.middleMethod();
    }

    middleMethod() {
        console.log("Middleman method delegating to another method");
        this.realMethod();
    }

    realMethod() {
        console.log("Real method doing the actual work");
    }

    accessInternalDetails() {
        // Inappropriate Intimacy (accessing internal details of another class)
        console.log(`Accessing internal details: ${this.chef.busy}`);
    }

    updateContactInfo(firstName, lastName, address, phoneNumber, email) {
        // Shotgun Surgery (changing multiple methods to update contact info)
        this.firstName = firstName;
        this.lastName = lastName;
        this.address = address;
        this.phoneNumber = phoneNumber;
        this.email = email;
    }

    updateName(firstName, lastName) {
        // Shotgun Surgery (separate method to update name)
        this.firstName = firstName;
        this.lastName = lastName;
    }

    updateAddress(address) {
        // Shotgun Surgery (separate method to update address)
        this.address = address;
    }

    updatePhoneNumber(phoneNumber) {
        // Shotgun Surgery (separate method to update phone number)
        this.phoneNumber = phoneNumber;
    }

    updateEmail(email) {
        // Shotgun Surgery (separate method to update email)
        this.email = email;
    }

    notifyForPromotion() {
        // Divergent Change (method to notify for promotion)
        console.log("Notifying customer for promotion");
    }

    notifyForDiscount() {
        // Divergent Change (method to notify for discount)
        console.log("Notifying customer for discount");
    }

    notifyForNewArrivals() {
        // Divergent Change (method to notify for new arrivals)
        console.log("Notifying customer for new arrivals");
    }

    applyDiscount(discountCode) {
        // Parallel Inheritance Hierarchies (method to apply discount)
        console.log(`Applying discount for customer with code ${discountCode}`);
    }

    applyLoyaltyPoints() {
        // Parallel Inheritance Hierarchies (method to apply loyalty points)
        console.log("Applying loyalty points for customer");
    }

    handleComplaint(complaint) {
        // Switch Statements (using if-else to simulate switch)
        if (complaint === "cold pizza") {
            this.calmCustomerDown();
        } else if (complaint === "late delivery") {
            this.calmCustomerDown();
        } else if (complaint === "wrong order") {
            this.calmCustomerDown();
        } else {
            this.calmCustomerDown();
        }
    }

    refusedBequest() {
        // Refused Bequest (method that should be inherited but is empty)
    }
}

module.exports = { Cashier };

File: C++/SmellyAnnotated/Cashier.cpp
/*
Copyright (c) 2025 Ahmed R. Sadik, Honda Research Institute Europe GmbH 

This source code is licensed under the MIT License found in the
LICENSE file in the root directory of this source tree. This dataset contains smelly code for research and refactoring purposes.
*/


#include "Cashier.h"

Cashier::Cashier(Chef* chef) : chef(chef), frequentCustomerDiscount(false) {}

void Cashier::takeOrder(const std::string& pizzaType) {
    std::cout << "Cashier is taking order for " << pizzaType << " pizza." << std::endl;
    chef->bakePizza(pizzaType);
}

void Cashier::hurryUpChef() {
    std::cout << "Cashier is hurrying up the chef." << std::endl;
    chef->hurryUp();
}

void Cashier::calmCustomerDown() {
    std::cout << "Cashier is calming the customer down." << std::endl;
}

void Cashier::deliverPizzaToCustomer() {
    std::cout << "Cashier is delivering pizza to the customer." << std::endl;
}

Chef* Cashier::getChef() {  // Implementation of getChef method
    return chef;
}

// Methods with code smells
void Cashier::notifyForPromotion() {
    std::cout << "Notifying customer for promotion" << std::endl;
}

void Cashier::notifyForDiscount() {
    std::cout << "Notifying customer for discount" << std::endl;
}

void Cashier::notifyForNewArrivals() {
    std::cout << "Notifying customer for new arrivals" << std::endl;
}

void Cashier::applyDiscount() {
    std::cout << "Applying discount for customer" << std::endl;
}

void Cashier::applyLoyaltyPoints() {
    std::cout << "Applying loyalty points for customer" << std::endl;
}

void Cashier::handleComplaint(const std::string& complaint) {
    if (complaint == "cold pizza") {
        std::cout << "Handling complaint: Pizza is cold" << std::endl;
    } else if (complaint == "late delivery") {
        std::cout << "Handling complaint: Pizza is late" << std::endl;
    } else if (complaint == "wrong order") {
        std::cout << "Handling complaint: Wrong pizza delivered" << std::endl;
    } else {
        std::cout << "Handling complaint: General complaint" << std::endl;
    }
}

void Cashier::askForReceipt() {
    std::cout << "Customer is asking for a receipt." << std::endl;
}

void Cashier::anotherUnusedMethod() {}

void Cashier::longMethod() {
    std::cout << "Cashier is handling many tasks in a single method" << std::endl;
    this->takeOrder("Cheese");
    this->notifyForPromotion();
    this->notifyForDiscount();
    this->notifyForNewArrivals();
    this->applyDiscount();
    this->applyLoyaltyPoints();
    this->handleComplaint("cold pizza");
}

void Cashier::orderWithUnnecessaryDetails(const std::string& pizzaType, const std::string& size, const std::string& crustType, const std::string& toppings, bool extraCheese, const std::string& discountCode) {
    std::cout << "Placing a detailed order for " << pizzaType << " pizza with " << size << ", " << crustType << ", " << toppings << ", extra cheese: " << extraCheese << ", discount code: " << discountCode << std::endl;
    this->takeOrder(pizzaType);
    this->applyDiscount();
    this->applyLoyaltyPoints();
}

void Cashier::duplicateMethod() {
    std::cout << "Cashier is making a duplicate order" << std::endl;
    this->takeOrder("Cheese");
    this->takeOrder("Cheese");
}

void Cashier::chainOfMethods() {
    std::cout << "Cashier is initiating a message chain" << std::endl;
    this->accessInternalDetails();
}

void Cashier::middlemanMethod() {
    std::cout << "Cashier is calling a middleman method" << std::endl;
    this->middleMethod();
}

void Cashier::middleMethod() {
    std::cout << "Middleman method delegating to another method" << std::endl;
    this->realMethod();
}

void Cashier::realMethod() {
    std::cout << "Real method doing the actual work" << std::endl;
}

void Cashier::accessInternalDetails() {
    // Inappropriate Intimacy
    std::cout << "Accessing internal details of Chef: " << chef->isBusy() << std::endl;
}

void Cashier::refusedBequest() {}
